WIP: Add python instructions to quickstart (#541)
MicrosoftDocs/quantum-docs-pr: articles/quickstart.md

---
title: Write your first quantum program
description: Learn how to write a quantum program in Q#. Develop a Bell State application using the Quantum Development Kit (QDK)
author: natke
ms.author: nakersha 
ms.date: 10/07/2019
ms.topic: tutorial
uid: microsoft.quantum.write-program
---
# Write your first quantum program

> Learn how to write a quantum program using the Microsoft QDK. You will write an application to demonstrate the simplest form of quantum entanglement, also known as **Bell State**.

Reviewer: ```suggestion
Learn how to write a quantum program using the Microsoft QDK. You will write an application to demonstrate the simplest form of quantum entanglement, a **Bell state**.
```

Reviewer: Feel free to play with the phrasing, but basically trying to avoid implying that the term "Bell state" is just a moniker for "the simplest form of quantum entanglement". Might be worth going with "one of the simplest forms..." as well, just to avoid being too concrete. 

(also, changed to "a" because there are actually four of them)

You start with the simplest program possible and build it up to demonstrate [quantum superposition](https://en.wikipedia.org/wiki/Quantum_superposition) and [quantum entanglement](https://en.wikipedia.org/wiki/Quantum_entanglement).

## Pre-requisites

* [Install](xref:microsoft.quantum.install) the Quantum Development Kit using your preferred language and development environment
* If you already have the QDK installed, make sure you have [updated](xref:microsoft.quantum.update) to the latest version

## What You'll Learn

> [!div class="checklist"]
> * How to set up a quantum solution and project
> * The components of a Q# operation
> * How to call a Q# operation from a host program

> > * How to create entanglement between two qubits

Author: I tried "How to entangle two qubits". What do you think?

## Setup

Applications developed with Microsoft's Quantum Development Kit consist of two parts:

1. One or more quantum algorithms, implemented using the Q# quantum programming language.
1. A host program, implemented in a programming language like Python or C# that serves as the main entry point and invokes Q# operations to execute a quantum algorithm.

#### [Python](#tab/tabid-python)

1. Choose a location for your application

1. Create a file called `Bell.qs`. This file will contain your Q# code.

> 1. Create a file called `host.py`. This file will contain your python host code.

Reviewer: ```suggestion
1. Create a file called `host.py`. This file will contain your Python host code.
```

#### [C# Command Line](#tab/tabid-csharp)

1. Create a new Q# project:

    ```bash
    dotnet new console -lang Q# --output Bell
    cd Bell
    ```

    You should see a `.csproj` file, a Q# file called `Operation.qs`, and a host program file called `Driver.cs`

1. Rename the Q# file

    ```bash
    mv Operation.qs Bell.qs
    ```

#### [Visual Studio](#tab/tabid-vs2019)

1. Create a new project

>    * Open Visual Studio

Reviewer 2: Some of these bullet points end in periods, while others don't? I'm not sure of the best style, though.

   * Go to the **File** menu and select **New** -> **Project...**
   * In the project template explorer, under `Installed` > `Visual C#`, select the `Q# Application` template
   * Make sure you have `.NET Framework 4.6.1` selected in the list at the bottom of the `New Project` dialog box.
   * Give your project the name `Bell`

1. Rename the Q# file

   * Navigate to the **Solution Explorer**
   * Right click on the `Operation.qs` file
   * Rename it to `Bell.qs`

* * *

## Write a Q# operation

> Our goal is to create a [Bell State](https://en.wikipedia.org/wiki/Bell_state) showing entanglement. We will build this up piece by piece to show the concepts of qubit state, gates, and measurement.

Reviewer 2: ```suggestion
Our goal is to prepare two qubits in a [Bell State](https://en.wikipedia.org/wiki/Bell_state), showing entanglement. We will build this up piece by piece to show the concepts of qubit state, gates, and measurement.
```

> Our goal is to create a [Bell State](https://en.wikipedia.org/wiki/Bell_state) showing entanglement. We will build this up piece by piece to show the concepts of qubit state, gates, and measurement.

Reviewer: ```suggestion
Our goal is to create a [Bell state](https://en.wikipedia.org/wiki/Bell_state) showing entanglement. We will build this up piece by piece to show the concepts of qubit state, gates, and measurement.
```

Reviewer: Unfortunately trying to get rid of the awkwardness of the "concept of qubit state," I added awkward alliteration. Maybe just "...of a qubit state, ..."?

Reviewer 3: maybe "...to demonstrate qubit states, gates, and measurement."

### Q# operation code

1. Replace the contents of the Bell.qs file with the following code:

    ```qsharp
    namespace Quantum.Bell {
        open Microsoft.Quantum.Intrinsic;
        open Microsoft.Quantum.Canon;

>         operation Set (desired: Result, q1: Qubit) : Unit {

Reviewer 2: ```suggestion
        operation Set(desired : Result, q1 : Qubit) : Unit {
```

            if (desired != M(q1)) {
                X(q1);
            }
        }
    }
    ```

    This operation may now be called to set a qubit to a known state (`Zero` or `One`).

>     We measure the qubit, if it's in the state we want, we leave it alone, otherwise, we flip it with the `X` gate.

Reviewer 2: Do we want `X` to be a link to the API reference for Microsoft.Quantum.Intrinsic.X?

Reviewer 3: I like this, but I'm not sure that API reference tells a new user anything too interesting beyond operation arguments? (which might be a sufficient amount of information)

### About Q# operations

An operation is the basic unit of quantum execution in Q#. It is roughly equivalent to a function in C or C++ or Python, or a static method in C# or Java.

The arguments to an operation are specified as a tuple, within parentheses.

> The return type of the operation is specified after a colon. In this case, the `Set` operation has no return, so it is marked as returning `Unit`. This is the Q# equivalent of `unit` in F#, which is roughly analogous to `void` in C#.

Reviewer 2: Would it be helpful to note that the analogous concept in Python is the empty tuple `()` of type `Tuple[()]`?

| | Value | Type |
|---|---|---|
| Q# | `()` | `Unit` |
| F# | `()` | `unit` | 
| Python | `()` | `Tuple[()]` |

### About quantum gates

You have used two quantum gates in your first Q# operation:

* The [M](xref:microsoft.quantum.intrinsic.m) gate, which measures the state of the qubit
* The [X](xref:microsoft.quantum.intrinsic.x) gate, which flips the state of a qubit

A quantum gate transforms the state of a qubit. The name comes from the analogy of a classical logic gate such as a `NOT` gate, which inverts the state of a classical bit.

### Add Q# test code

1. Add the following operation to the `Bell.qs` file, inside the namespace, after the end of the
`Set` operation:

    ```qsharp

>     operation BellTest (count : Int, initial: Result) : (Int, Int) {

Reviewer 2: I'd suggest that this is a good opportunity to rename operations and functions in the documentation to better align with the style guide. For instance, in the original quickstart, the name `BellTest` isn't a verb, such that it's not clear that it has a side effect on the quantum device. Especially since there's a loosely related other concept in quantum computing known as a "Bell test," (see the CHSH game sample), perhaps we could rename to `TestBellStatePreparation` or something similar?

---
title: Write your first quantum program
description: Learn how to write a quantum program in Q#. Develop a Bell State application using the Quantum Development Kit (QDK)
author: natke
ms.author: nakersha 
ms.date: 10/07/2019
ms.topic: tutorial
uid: microsoft.quantum.write-program
---
# Write your first quantum program

> Learn how to write a quantum program using the Microsoft QDK. You will write an application to demonstrate quantum entanglement using a **Bell State**.

Reviewer: ```suggestion
Learn how to write a quantum program using the Microsoft QDK. You will write an application to demonstrate quantum entanglement using a **Bell state**.
```

   * Navigate to the **Solution Explorer**
   * Right click on the `Operation.qs` file
   * Rename it to `Bell.qs`

* * *

## Write a Q# operation

Our goal is to prepare two qubits in a [Bell state](https://en.wikipedia.org/wiki/Bell_state) showing entanglement. We will build this up piece by piece to demonstrate qubit states, gates, and measurement.

### Q# operation code

1. Replace the contents of the Bell.qs file with the following code:

    ```qsharp
    namespace Quantum.Bell {
        open Microsoft.Quantum.Intrinsic;
        open Microsoft.Quantum.Canon;

        operation Set(desired : Result, q1 : Qubit) : Unit {
            if (desired != M(q1)) {
                X(q1);
            }
        }
    }
    ```

    This operation may now be called to set a qubit to a known state (`Zero` or `One`).

    We measure the qubit, if it's in the state we want, we leave it alone, otherwise, we flip it with the `X` gate.

### About Q# operations

An operation is the basic unit of quantum execution in Q#. It is roughly equivalent to a function in C or C++ or Python, or a static method in C# or Java.

The arguments to an operation are specified as a tuple, within parentheses.

> The return type of the operation is specified after a colon. In this case, the `Set` operation has no return, so it is marked as returning `Unit`. This is the Q# equivalent of `unit` in F#, which is roughly analogous to `void` in C#, and an empty tuple (`Tuple[()]`) in Python.

Reviewer 2: I'll readily admit that this is hair splitting, so please feel free to disregard, but all Q# operations return something, even if that is a container with no contents (i.e.: the empty tuple of type `Unit`).

        // Return number of times we saw a |0> and number of times we saw a |1>
        return (count-numOnes, numOnes);
    }
    ```

    This operation (`TestBellState`) will loop for `count` iterations, set a specified `initial` value on a qubit and then measure (`M`) the result. It will gather statistics on how many zeros and ones we've measured and return them to the caller. It performs one other necessary operation. It resets the qubit to a known state (`Zero`) before returning it allowing others to allocate this qubit in a known state. This is required by the `using` statement.

### About variables in Q#

Q# deals with variables in a unique way. By default, variables in Q# are immutable; their value may not be changed after they are bound. The `let` keyword is used to indicate the binding of an immutable variable. Operation arguments are always immutable.

If you need a variable whose value can change, such as `numOnes` in the example, you can declare the variable with the `mutable` keyword. A mutable variable's value may be changed using a `set` statement.

> In both cases, the type of a variable is inferred by the compiler. Q# doesn't require any type annotations for variables.

Reviewer 2: ```suggestion
In both cases, the type of a variable is inferred by the compiler. Q# doesn't require (or allow) any type annotations for variables.
```

    ```csharp
    using System;

    using Microsoft.Quantum.Simulation.Core;
    using Microsoft.Quantum.Simulation.Simulators;

    namespace Quantum.Bell
    {
        class Driver
        {
            static void Main(string[] args)
            {
                using (var qsim = new QuantumSimulator())
                {
                    // Try initial values
                    Result[] initials = new Result[] { Result.Zero, Result.One };
                    foreach (Result initial in initials)
                    {
                        var res = BellTest.Run(qsim, 1000, initial).Result;
                        var (numZeros, numOnes) = res;
                        System.Console.WriteLine(
                            $"Init:{initial,-4} 0s={numZeros,-4} 1s={numOnes,-4}");
                    }
                }

                System.Console.WriteLine("Press any key to continue...");
                Console.ReadKey();
            }
        }
    }
    ```

#### [](#tab/tabid-vs2019)

* * *

### About the host application code

#### [Python](#tab/tabid-python)

The Python host application has three parts:

> * Compute any arguments required for the quantum algorithm. In the example, `count` is fixed at a 1000 and `initial` is the initial value of the qubit.

Reviewer: This was actually another nuance I forgot to mention in the call yesterday, that might interest @cgranade too. By way of fumbling and bumbling I found that putting "0" and "1" in as the initial values in the Python call works. I learned the hard way after trying the "Result.Zero"/"Result.One" without knowing to import "Result" from qsharp as you did here--so thank you, I'll change to that in my tutorial.

Chris, in particular: during that fumbling, I learned that any input like "0.5" or "2" also worked without throwing any errors (though the output wasn't what I expected). I didn't look for a pattern or anything, so I'm not sure if it's a bug or something weird but expected, but figured I'd give you a heads up anyhow.

Reviewer 2: @gillenhaalb: Would you be willing to file that as a bug on microsoft/iqsharp? That's good to know, thank you!

Reviewer: Just put it up, no problem!

Reviewer 2: Thanks! Resolving the conversation here, then, as I don't think there's anything further that can be actioned on that from within this PR.

The program will exit after you press a key.

* * *

> ## Create superposition

Reviewer 2: ```suggestion
## Prepare superposition
```

Now we want to manipulate the qubit. First we'll just try to flip it. This is accomplished by performing an `X` gate before we measure it in `BellTest`:

```qsharp
X(qubit);

> let res = M (qubit);

Reviewer 2: ```suggestion
let res = M(qubit);
```

However, everything we've seen so far is classical. Let's get a quantum result. All we need to do is replace the `X` gate in the previous run with an `H` or Hadamard gate. Instead of flipping the qubit all the way from 0 to 1, we will only flip it halfway. The replaced lines in `BellTest` now look like:

```qsharp
H(qubit);

> let res = M (qubit);

Reviewer 2: ```suggestion
let res = M(qubit);
```

Every time we measure, we ask for a classical value, but the qubit is halfway between 0 and 1, so we get (statistically) 0 half the time and 1 half the time. This is known as __superposition__ and gives us our first real view into a quantum state.

> ## Create entanglement

Reviewer 2: ```suggestion
## Prepare entanglement
```

> Now we'll make the promised [Bell State](https://en.wikipedia.org/wiki/Bell_state) and show off __entanglement__. The first thing we'll need to do is allocate 2 qubits instead of one in `BellTest`:

Reviewer: ```suggestion
Now we'll make the promised [Bell state](https://en.wikipedia.org/wiki/Bell_state) and show off __entanglement__. The first thing we'll need to do is allocate 2 qubits instead of one in `BellTest`:
```

```qsharp
using ((q0, q1) = (Qubit(), Qubit())) {
```

This will allow us to add a new gate (`CNOT`) before we measure  (`M`) in `BellTest`:

```qsharp

> Set (initial, q0);

Reviewer 2: ```suggestion
Set(initial, q0);
```

> Set (Zero, q1);

Reviewer 2: ```suggestion
Set(Zero, q1);
```

H(q0);
CNOT(q0,q1);

> let res = M (q0);

Reviewer 2: ```suggestion
let res = M(q0);
```

Our statistics for the first qubit haven't changed (50-50 chance of a 0 or a 1), but now when we measure the second qubit, it is __always__ the same as what we measured for the first qubit. Our `CNOT` has entangled the two qubits, so that whatever happens to one of them, happens to the other. If you reversed the measurements (did the second qubit before the first), the same thing would happen. The first measurement would be random and the second would be in lock step with whatever was discovered for the first.

Congratulations, you've written your first quantum program!

> 

Reviewer 2: I totally agree with removing this from the current quickstart, but it'd be good if we could land it somewhere else instead if at all possible.

Reviewer: Perhaps a future tutorial demonstrating the machines beyond the full state simulator? Would probably fit perfectly to a single tutorial if we just provide a bare bones situation of when each machine would be useful, and then quickly work through running on it and examining the output.

Author: Yes, I do agree with this, and we should definitely use the material in another tutorial or how to

## What's next?

For more information about the metrics reported and accessing the data programmatically,
take a look at the [`ResourcesEstimator` documentation](xref:microsoft.quantum.machines.resources-estimator).

> To learn more about the other type of simulators and target machines provided in the Quantum Development Kit, how they work and how to use them, take a look at the [Managing Quantum machines and drivers topic](xref:microsoft.quantum.machines) in the documentation.

Reviewer: ```suggestion
To learn more about the other type of simulators and target machines provided in the Quantum Development Kit, how they work and how to use them, take a look at the [Managing quantum machines and drivers topic](xref:microsoft.quantum.machines) in the documentation.
```
